Fix WebSocket `_onWSClose` handler: do not ignore close codes

Akshitguptaa opened the pull request:
## References

Fixes #18189

## Code changes

<!-- Describe the code changes and how they address the issue. -->
Modified the `_onWSClose` handler in both `packages/services/src/kernel/default.ts` and `packages/services/src/terminal/default.ts`.

Previously, these handlers blindly attempted to reconnect on *any* WebSocket close event. This caused unnecessary reconnection loops even when the connection was closed gracefully (e.g., during a manual kernel shutdown).

**Changes implemented:**
* Updated the `_onWSClose` signature to accept `CloseEvent` instead of generic `Event` to access the `.code` property.
* Added a check for standard close codes `1000` (Normal Closure) and `1001` (Going Away).
* If a normal close code is detected, the handler now correctly stops the reconnection process.

## User-facing changes

<!-- Describe any visual or user interaction changes and how they address the issue. -->
Users will no longer experience unnecessary reconnection attempts (or infinite "Reconnecting..." loops) when a kernel or terminal is gracefully shut down.

<!-- For visual changes, include before and after screenshots or GIF/mp4/other video demo here. -->
**Debugger Screenshots** 
https://postimg.cc/dhCh0FRw
<!-- https://postimg.cc/gxbYNZhT -->

## Backwards-incompatible changes

<!-- Describe any backwards-incompatible changes to JupyterLab public APIs. -->
None.

Akshitguptaa committed:
Fix: Handle WebSocket close codes 1000/1001 in Kernel and Terminal

jianyexi commented:
@Akshitguptaa thanks for fixing this issue, can we merge this to unblock our component

krassowski commented on packages/services/src/kernel/default.ts:
```suggestion
    const earlyClose = async (evt: CloseEvent | ErrorEvent) => {
```

krassowski reviewed (commented):
A minor note on type correctness; for the operations, did you test the behavior with various kinds of kernel restarts and network issues? I know that it should work, but I the error codes sent are not always what they should be so just wanted to know how much testing this had.

Akshitguptaa committed:
Update packages/services/src/kernel/default.ts

Co-authored-by: Michał Krassowski <5832902+krassowski@users.noreply.github.com>

Akshitguptaa committed:
Update packages/services/src/kernel/default.ts

Co-authored-by: Michał Krassowski <5832902+krassowski@users.noreply.github.com>

Akshitguptaa committed:
Update packages/services/src/kernel/default.ts

Co-authored-by: Michał Krassowski <5832902+krassowski@users.noreply.github.com>

Akshitguptaa committed:
Update packages/services/src/kernel/default.ts

Co-authored-by: Michał Krassowski <5832902+krassowski@users.noreply.github.com>

Akshitguptaa committed:
Merge branch 'jupyterlab:main' into fix/websocket-close-codes

Akshitguptaa committed:
fix: add type guard

Akshitguptaa commented:
@krassowski  Thanks! I applied the type suggestions. 
I also verified locally that manual shutdowns, kernel restarts, and network drops all behave exactly as expected now without throwing any console errors.

Commit:
fix: use property check instead of instanceof for CloseEvent and add unit tests

- Replace 'evt instanceof CloseEvent' with '"code" in evt' because
  CloseEvent is not a global in Node.js, causing the check to always
  fail in tests (the ws library creates its own internal CloseEvent).
- Add closeWithCode() helper to SocketTester for close code testing.
- Add ws_close_codes.spec.ts with 4 tests:
  - auto-reconnect on default close (no code)
  - no reconnect on normal closure (code 1000)
  - no reconnect on going away (code 1001)
  - reconnect on abnormal closure (code 1011)

Akshitguptaa committed:
style: run prettier on ws_close_codes.spec.ts

jianyexi commented:
@krassowski Gentle ping — the unit tests and reproduction steps you requested have been added (commit cherry-picked by @Akshitguptaa). The PR now includes 4 tests covering close codes 1000, 1001, 1011, and default close. Also fixed an \instanceof CloseEvent\ issue that didn't work in Node.js tests. Would you be able to take a look when you get a chance? Thanks!

krassowski reviewed (approved):
Thank you, let's try it.
